src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -286,52 +439,119 @@ static size_t uct_cuda_copy_md_get_total_device_mem(CUdevice cuda_device)
     return 1; /* return 1 byte to avoid division by zero */
 }
 
+static ucs_status_t
+uct_cuda_copy_detect_vmm(const void *address, CUmemorytype *cuda_mem_type,
+                         unsigned *is_vmm_allocation)
+{
+    ucs_status_t status = UCS_OK;
+#if HAVE_CUDA_FABRIC
+    CUmemAllocationProp prop = {};
+    CUmemGenericAllocationHandle alloc_handle;
+
+    /* Check if memory is allocated using VMM API and see if host memory needs
+     * to be treated as pinned device memory */
+    status = UCT_CUDADRV_FUNC_LOG_DEBUG(
+            cuMemRetainAllocationHandle(&alloc_handle, (void*)address));
```

Comment:
why need to "retain"?

Reply:
This is the only API to get a reference to allocation handle backing the range. We retain and release in the same function. Not sure if there is a concern.